README.md
# stactools-noaa-c-cap

- Name: noaa-c-cap
- Package: `stactools.noaa_c_cap`
- PyPI: https://pypi.org/project/stactools-noaa-c-cap/
- Owner: @gadomski
- Dataset homepage: http://github.com/stactools-packages/noaa-c-cap
- STAC extensions used:
  - [classification](https://github.com/stac-extensions/classification)
  - [item-assets](https://github.com/stac-extensions/item-assets)
  - [proj](https://github.com/stac-extensions/projection)
  - [raster](https://github.com/stac-extensions/raster)
  - [scientific](https://github.com/stac-extensions/scientific)
- Extra fields: None

Create STAC Items and Collections for NOAA C-CAP data.

[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/stactools-packages/noaa-c-cap/main?filepath=docs/installation_and_basic_usage.ipynb)

## Examples

### STAC objects

- [Collection](examples/collection.json)
- [Item](examples/conus_2016_ccap_landcover_20200311/conus_2016_ccap_landcover_20200311.json)

### Command-line usage

Create a single item

```bash
$ stac noaa-c-cap create-item conus_2016_ccap_landcover_20200311.tif item.json
```

Create the entire collection from the remote data sources:

```bash
$ stac noaa-c-ccap create-collection destination-directory
```

Creating the `examples/` directory

```bash
$ stac noaa-c-cap create-collection --catalog-type SELF_CONTAINED examples
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). This project attempts to match the major and minor versions of [stactools](https://github.com/stac-utils/stactools) and increments the patch number as needed.

## [Unreleased]

## [v0.2.2] - 2022-05-18

### Added

- License link, boundary asset, and more extents ([#17](https://github.com/stactools-packages/noaa-c-cap/pull/17))

## [v0.2.1] - 2022-04-19

### Fixed

- Removed leftover `gsd` attribute -- it is now in `raster:bands` ([#15](https://github.com/stactools-packages/noaa-c-cap/pull/15))

## [v0.2.0] - 2022-04-18

### Added

- `classification` extension ([#12](https://github.com/stactools-packages/noaa-c-cap/pull/12))
- `raster` extension and links ([#13](https://github.com/stactools-packages/noaa-c-cap/pull/13))
- pre-commit ([#13](https://github.com/stactools-packages/noaa-c-cap/pull/13))

### Changed

- All in [#12](https://github.com/stactools-packages/noaa-c-cap/pull/12):
    - Use `pytest` instead of `unittest`
    - Use `black` instead of `yapf`
    - Set datetime to `None`

### Removed

- `file` and `label` extensions ([#12](https://github.com/stactools-packages/noaa-c-cap/pull/12))

## [0.1.4] - 2021-01-19

### Added

- `file` extension

## [0.1.3] - 2022-01-03

### Added

- `read_href_modifier` to `create_item`

## [0.1.2] - 2021-09-22

### Added

- Ability to COGify tiffs while downloading
- Item Assets extension to the Collection
- GSD to Items and Collections
- Constant label classes
- Scientific extension for a citation on the Collection

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## [0.1.1] - 2021-09-20

### Added

- Nothing

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Lints.

## [0.1.0] - 2021-09-20

Initial commit.

### Added

- Everything.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

[Unreleased]: <https://github.com/stactools-packages/noaa-c-cap/compare/v0.2.2..main>
[0.2.2]: <https://github.com/stactools-packages/noaa-c-cap/compare/v0.2.1..v0.2.2>
[0.2.1]: <https://github.com/stactools-packages/noaa-c-cap/compare/v0.2.0..v0.2.1>
[0.2.0]: <https://github.com/stactools-packages/noaa-c-cap/compare/v0.1.4..v0.2.0>
[0.1.4]: <https://github.com/stactools-packages/noaa-c-cap/compare/v0.1.3..v0.1.4>
[0.1.3]: <https://github.com/stactools-packages/noaa-c-cap/compare/v0.1.2..v0.1.3>
[0.1.2]: <https://github.com/stactools-packages/noaa-c-cap/compare/v0.1.1..v0.1.2>
[0.1.1]: <https://github.com/stactools-packages/noaa-c-cap/compare/v0.1.0..v0.1.1>
[0.1.0]: <https://github.com/stactools-packages/noaa-c-cap/releases/tag/v0.1.0>


tests/__init__.py
from stactools.testing.test_data import TestData

EXTERNAL_DATA = {
    "conus_2016_ccap_landcover_20200311.tif": {
        "url": (
            "https://coast.noaa.gov/htdata/raster1/landcover/"
            "bulkdownload/30m_lc/conus_2016_ccap_landcover_20200311.tif"
        ),
        "compress": None,
    },
    "conus_2010_ccap_landcover_20200311.tif": {
        "url": (
            "https://coast.noaa.gov/htdata/raster1/landcover/"
            "bulkdownload/30m_lc/conus_2010_ccap_landcover_20200311.tif"
        ),
        "compress": None,
    },
    "pr_2010_ccap_hr_land_cover20170214_30m.tif": {
        "url": (
            "https://coast.noaa.gov/htdata/raster1/landcover/"
            "bulkdownload/30m_lc/pr_2010_ccap_hr_land_cover20170214_30m.tif"
        ),
        "compress": None,
    },
}

test_data = TestData(__file__, external_data=EXTERNAL_DATA)


tests/test_stac.py
import datetime
import os.path
import unittest
from tempfile import TemporaryDirectory

from pystac import MediaType
from pystac.extensions.item_assets import ItemAssetsExtension
from pystac.extensions.scientific import ScientificExtension

from stactools.noaa_c_cap import stac
from stactools.noaa_c_cap.constants import CLASSIFICATION_CLASSES, COLLECTION_CITATION
from tests import test_data


class StacTest(unittest.TestCase):
    def test_create_collection(self):
        hrefs = [
            test_data.get_external_data("conus_2016_ccap_landcover_20200311.tif"),
            test_data.get_path("data-files/CCAP_Parent_2016.xml"),
        ]
        collection = stac.create_collection(hrefs)
        self.assertEqual(collection.id, "noaa-c-cap")
        self.assertEqual(collection.title, "C-CAP Regional Land Cover and Change")
        self.assertEqual(len(list(collection.get_all_items())), 1)

        item_assets = ItemAssetsExtension.ext(collection)
        assert "data" in item_assets.item_assets
        assert "metadata" in item_assets.item_assets

        scientific = ScientificExtension.ext(collection)
        self.assertEqual(scientific.citation, COLLECTION_CITATION)

        assert (
            "https://stac-extensions.github.io/classification/v1.0.0/schema.json"
            in collection.stac_extensions
        )

        with TemporaryDirectory() as directory:
            collection.normalize_hrefs(os.path.join(directory, "collection.json"))
            collection.validate_all()

    def test_create_item(self):
        path = test_data.get_external_data("conus_2016_ccap_landcover_20200311.tif")
        item = stac.create_item(path)
        self.assertEqual(item.id, "conus_2016_ccap_landcover_20200311")
        assert item.datetime is None
        self.assertEqual(
            item.common_metadata.start_datetime,
            datetime.datetime(2016, 1, 1, tzinfo=datetime.timezone.utc),
        )
        self.assertEqual(
            item.common_metadata.end_datetime,
            datetime.datetime(2016, 12, 31, tzinfo=datetime.timezone.utc),
        )
        self.assertIsNone(item.common_metadata.gsd)
        data = item.assets["data"]
        self.assertEqual(data.media_type, MediaType.COG)
        self.assertEqual(data.roles, ["data"])

        assert (
            "https://stac-extensions.github.io/classification/v1.0.0/schema.json"
            in item.stac_extensions
        )
        assert data.extra_fields["classification:classes"] == CLASSIFICATION_CLASSES

        item.validate()

    def test_create_item_with_read_href_modifier(self):
        did_it = False

        def do_it(href: str) -> str:
            nonlocal did_it
            did_it = True
            return href

        path = test_data.get_external_data("conus_2016_ccap_landcover_20200311.tif")
        stac.create_item(path, read_href_modifier=do_it)
        assert did_it

    def test_create_item_with_xml(self):
        tiff_path = test_data.get_external_data(
            "conus_2016_ccap_landcover_20200311.tif"
        )
        xml_path = test_data.get_path("data-files/CCAP_Parent_2016.xml")
        item = stac.create_item(tiff_path, xml_path)
        metadata = item.assets["metadata"]
        self.assertEqual(metadata.media_type, MediaType.XML)
        self.assertEqual(metadata.roles, ["metadata"])
        item.validate()

    def test_create_item_2010(self):
        tiff_path = test_data.get_external_data(
            "conus_2016_ccap_landcover_20200311.tif"
        )
        xml_path = test_data.get_path("data-files/CCAP_Parent_2010.xml")
        item = stac.create_item(tiff_path, xml_path)
        item.validate()

    def test_create_item_puerto_rico(self):
        tiff_path = test_data.get_external_data(
            "pr_2010_ccap_hr_land_cover20170214_30m.tif"
        )
        xml_path = test_data.get_path("data-files/2010_puerto_rico_ccap.xml")
        item = stac.create_item(tiff_path, xml_path)
        item.validate()


tests/test_module.py
import unittest

import stactools.noaa_c_cap


class TestModule(unittest.TestCase):
    def test_version(self):
        self.assertIsNotNone(stactools.noaa_c_cap.__version__)


src/stactools/noaa_c_cap/stac.py
import datetime
import logging
from typing import List, Optional

from pystac import Asset, Collection, Extent, Item, MediaType, Summaries
from pystac.extensions.item_assets import AssetDefinition, ItemAssetsExtension
from pystac.extensions.raster import DataType, RasterBand, RasterExtension
from pystac.extensions.scientific import ScientificExtension
from stactools.core import create
from stactools.core.io import ReadHrefModifier

from stactools.noaa_c_cap import Dataset, utils
from stactools.noaa_c_cap.constants import (
    CLASSIFICATION_CLASSES,
    CLASSIFICATION_EXTENSION_HREF,
    COLLECTION_CITATION,
    COLLECTION_DESCRIPTION,
    COLLECTION_ID,
    COLLECTION_KEYWORDS,
    COLLECTION_LINKS,
    COLLECTION_PROVIDERS,
    COLLECTION_TITLE,
    SPATIAL_RESOLUTION,
)

logger = logging.getLogger(__name__)


def create_collection(hrefs: Optional[List[str]] = None) -> Collection:
    """Creates a STAC Collection for all C-CAP data.

    Args:
        hrefs (Optional[str]): If provided, create a Collection for the provided
        HREFs. If not provided, use the network locations defined in
        `urllist-modres.txt`.
    """
    if not hrefs:
        hrefs = utils.urls()
    datasets = Dataset.from_hrefs(hrefs)
    items = [create_item_from_dataset(dataset) for dataset in datasets]
    extent = Extent.from_items(items)
    for prefix in ("conus", "hi", "pr"):
        secondary_items = [item for item in items if item.id.startswith(prefix)]
        if secondary_items:
            secondary_extent = Extent.from_items(secondary_items)
            extent.spatial.bboxes.extend(secondary_extent.spatial.bboxes)
            extent.temporal.intervals.extend(
                secondary_extent.temporal.intervals  # type: ignore
            )
    summaries = Summaries.empty()
    collection = Collection(
        id=COLLECTION_ID,
        title=COLLECTION_TITLE,
        description=COLLECTION_DESCRIPTION,
        extent=extent,
        keywords=COLLECTION_KEYWORDS,
        providers=COLLECTION_PROVIDERS,
        summaries=summaries,
    )
    collection.add_links(COLLECTION_LINKS)
    collection.add_items(items)
    collection.add_asset(
        "boundary",
        Asset(
            href="https://coast.noaa.gov/data/digitalcoast/zip/ccap-mapping-bndry-wgs84.zip",
            title="C-CAP Mapping Boundary (shp file)",
            media_type="application/zip",
        ),
    )

    item_assets = {}
    for item in items:
        for key, asset in item.get_assets().items():
            asset_as_dict = asset.to_dict()
            asset_as_dict.pop("href")
            if key not in item_assets:
                item_assets[key] = AssetDefinition(asset_as_dict)
            elif item_assets[key].to_dict() != asset_as_dict:
                logger.warning(
                    f"Item Asset '{key}' does not match asset:\n"
                    f"item_asset={item_assets[key].to_dict()}\n"
                    f"asset={asset_as_dict}"
                )

    item_assets_ext = ItemAssetsExtension.ext(collection, add_if_missing=True)
    item_assets_ext.item_assets = item_assets

    scientific = ScientificExtension.ext(collection, add_if_missing=True)
    scientific.citation = COLLECTION_CITATION

    RasterExtension.add_to(collection)
    collection.stac_extensions.append(CLASSIFICATION_EXTENSION_HREF)

    return collection


def create_item(
    tiff_href: str,
    xml_href: Optional[str] = None,
    read_href_modifier: Optional[ReadHrefModifier] = None,
) -> Item:
    """Creates a STAC Item for the provided C-CAP tiff file.

    Args:
        tiff_href (str): The HREF pointing to the tiff file.
        xml_href (str): The HREF pointing to the xml file.
        read_href_modifier (Callable[[str], str]): A function to modify the read href for the file.

    Returns:
        Item: STAC Item object
    """
    return create_item_from_dataset(
        Dataset(tiff_href=tiff_href, xml_href=xml_href),
        read_href_modifier=read_href_modifier,
    )


def create_item_from_dataset(
    dataset: Dataset, read_href_modifier: Optional[ReadHrefModifier] = None
) -> Item:
    """Creates an item from a NOAA C-CAP dataset."""
    logger.info(f"Creating STAC item from {dataset.tiff_href}")
    item = create.item(dataset.tiff_href, read_href_modifier=read_href_modifier)
    item.datetime = None
    item.common_metadata.start_datetime = datetime.datetime(
        int(dataset.year), 1, 1, tzinfo=datetime.timezone.utc
    )
    item.common_metadata.end_datetime = datetime.datetime(
        int(dataset.year), 12, 31, tzinfo=datetime.timezone.utc
    )

    data = item.assets.get("data")
    assert data
    data.media_type = MediaType.COG

    raster = RasterExtension.ext(data, add_if_missing=True)
    raster.bands = [
        RasterBand.create(
            data_type=DataType.UINT8,
            spatial_resolution=SPATIAL_RESOLUTION,
        )
    ]

    item.stac_extensions.append(CLASSIFICATION_EXTENSION_HREF)
    data.extra_fields["classification:classes"] = CLASSIFICATION_CLASSES

    if dataset.xml_href:
        item.add_asset(
            "metadata",
            Asset(dataset.xml_href, media_type=MediaType.XML, roles=["metadata"]),
        )
    return item


src/stactools/noaa_c_cap/dataset.py
import logging
import os.path
from typing import List, Optional

logger = logging.getLogger(__name__)


class Dataset:
    """A NOAA C-CAP dataset.

    Could be CONUS, could be somewhere else. Always has a tiff file, can have XML metadata.
    """

    tiff_href: str
    year: str
    location: str
    xml_file_name: Optional[str]
    xml_href: Optional[str]
    classes: List[str]

    @classmethod
    def from_hrefs(cls, hrefs: List[str]) -> List["Dataset"]:
        """Creates one or more datasets from a list of hrefs.

        Does the work of looking through the HREFs to associate tifs with xml metadata.
        """
        datasets = []
        for tif_href in (href for href in hrefs if os.path.splitext(href)[1] == ".tif"):
            dataset = Dataset(tif_href)
            if dataset.xml_file_name:
                xml_href = next(
                    (
                        href
                        for href in hrefs
                        if os.path.basename(href) == dataset.xml_file_name
                    ),
                    None,
                )
                if xml_href:
                    dataset.xml_href = xml_href
                else:
                    logger.warn(f"Could not find XML metadata file for {tif_href}")
            datasets.append(dataset)
        return datasets

    def __init__(self, tiff_href: str, xml_href: Optional[str] = None):
        """Creates a new dataset for a GeoTIFF."""
        self.tiff_href = tiff_href
        self.xml_href = xml_href

        file_name = os.path.basename(tiff_href)
        parts = file_name.split("_")
        if len(parts) < 2:
            raise ValueError(f"Invalid NOAA C-CAP file name: {file_name}")
        self.location = parts[0]
        self.year = parts[1]
        if self.location == "conus":
            self.xml_file_name: Optional[str] = f"CCAP_Parent_{self.year}.xml"
        elif self.location == "hi":
            self.xml_file_name = None
        elif self.location == "pr":
            self.xml_file_name = f"{self.year}_puerto_rico_ccap.xml"


src/stactools/noaa_c_cap/commands.py
import logging
import os
from tempfile import TemporaryDirectory
from typing import Optional, Tuple

import click
import requests
import stactools.core.utils.convert
from click.types import Choice
from pystac.catalog import CatalogType

from stactools.noaa_c_cap import stac, utils

logger = logging.getLogger(__name__)


def create_noaa_c_cap_command(cli):
    """Creates the stactools-noaa-c-cap command line utility."""

    @cli.group(
        "noaa-c-cap",
        short_help=("Commands for working with stactools-noaa-c-cap"),
    )
    def noaa_c_cap():
        pass

    @noaa_c_cap.command(
        "create-collection",
        short_help="Creates a STAC collection",
    )
    @click.argument("destination")
    @click.option("--href", "-h", multiple=True)
    @click.option("--directory", "-d", help="Directory containing NOAA C-CAP files")
    @click.option(
        "--catalog-type",
        type=Choice(["ABSOLUTE_PUBLISHED", "RELATIVE_PUBLISHED", "SELF_CONTAINED"]),
        default="ABSOLUTE_PUBLISHED",
    )
    def create_collection_command(
        destination: str,
        href: Tuple[str],
        directory: Optional[str],
        catalog_type: CatalogType,
    ):
        """Creates a STAC Collection
        Args:
            destination (str): An HREF for the Collection JSON
        """
        hrefs = list(href)
        if directory:
            hrefs.extend(
                os.path.join(os.path.abspath(directory), file_name)
                for file_name in os.listdir(directory)
            )
        collection = stac.create_collection(hrefs=hrefs)
        collection.catalog_type = catalog_type
        collection.normalize_and_save(destination)

    @noaa_c_cap.command("create-item", short_help="Create a STAC item")
    @click.argument("source")
    @click.argument("destination")
    @click.option("-x", "--xml", help="HREF to the metadata XML file.")
    @click.option(
        "--cogify/--no-cogify",
        default=False,
        help=(
            "Use a Cloud-Optimized GeoTIFF version "
            "of the input file as the data asset."
        ),
    )
    def create_item_command(
        source: str, destination: str, xml: Optional[str], cogify: bool
    ):
        """Creates a STAC Item

        Args:
            source (str): HREF of the Asset associated with the Item
            destination (str): An HREF for the STAC Collection
            xml (str): HREF to the metadata XML file
            cogify (bool): Use a Cloud-Optimized GeoTIFF version of the source
            file as the data asset
        """
        if cogify:
            cog = os.path.join(os.path.dirname(destination), os.path.basename(source))
            stactools.core.utils.convert.cogify(source, cog)
            source = cog
        item = stac.create_item(source, xml)
        item.save_object(dest_href=destination)

    @noaa_c_cap.command(
        "download", short_help="Download NOAA C-CAP data into a target directory"
    )
    @click.argument("destination")
    @click.option(
        "--cogify/--no-cogify",
        help="Convert the .tifs to COGs after download",
        default=False,
    )
    def download_command(destination: str, cogify: bool) -> None:
        """Downloads the source data to a target directory.

        Args:
            destination (str): A local directory into which the data will downloaded.
        """
        os.makedirs(destination, exist_ok=True)
        with TemporaryDirectory() as temporary_directory:
            for url in utils.urls():
                file_name = os.path.basename(url)
                final_path = os.path.join(destination, file_name)
                print(f"Downloading {url} to {final_path}")
                download_to_temporary_directory = (
                    cogify and os.path.splitext(file_name)[1] == ".tif"
                )
                if download_to_temporary_directory:
                    download_path = os.path.join(temporary_directory, file_name)
                else:
                    download_path = final_path
                with open(download_path, "wb") as f:
                    response = requests.get(url, stream=True)
                    for chunk in response.iter_content(chunk_size=1024):
                        f.write(chunk)
                if download_to_temporary_directory:
                    stactools.core.utils.convert.cogify(download_path, final_path)

    return noaa_c_cap


src/stactools/noaa_c_cap/metadata.py
import datetime
from typing import List
from xml.etree import ElementTree
from xml.etree.ElementTree import Element

import fsspec


class Metadata:
    publication_date: datetime.datetime
    classes: List[str]

    def __init__(self, href: str):
        with fsspec.open(href) as file:
            string = file.read()
        xml = ElementTree.fromstring(string)
        publication_date = xml.findtext('item-identification/publication-date')
        if publication_date:
            self.publication_date = datetime.datetime.strptime(
                publication_date,
                r'%Y-%m-%d').replace(tzinfo=datetime.timezone.utc)
        else:
            timeperd = xml.findtext(
                'idinfo/timeperd/timeinfo/sngdate/caldate')  # puerto rico
            if timeperd:
                self.publication_date = datetime.datetime.strptime(
                    timeperd, r'%Y%m%d')
        # 2016 CONUS
        allowed_values = xml.findtext(
            'entity-attribute-information/entity/data-attributes/data-attribute/allowed-values'
        )
        # Puerto Rico
        eainfo = xml.find('eainfo')
        # Default
        supplemental_information = xml.findtext(
            'item-identification/supplemental-information')
        if supplemental_information:
            self.classes = parse_supplemental_information(
                supplemental_information)
        elif eainfo:
            self.classes = parse_eainfo(eainfo)
        elif allowed_values:
            self.classes = parse_allowed_values(allowed_values)


def parse_allowed_values(text: str) -> List[str]:
    classes = ['Unclassified']
    for entry in text.split(' | '):
        values = entry.split(': ')
        assert len(values) == 2
        assert int(values[0]) == len(classes)
        classes.append(values[1])
    return classes


def parse_eainfo(eainfo: Element) -> List[str]:
    classes = ['Unclassified']
    for element in eainfo.findall('detailed/attr/attrdomv'):
        text = element.findtext('edom/edomv')
        assert text
        parts = text.split()
        assert (len(parts) >= 2)
        value = int(parts[0])
        assert len(classes) == value
        classes.append(' '.join(parts[1:]))
    return classes


def parse_supplemental_information(text: str) -> List[str]:
    lines = text.splitlines()
    assert (len(lines) > 1)
    classes: List[str] = []
    for line in lines[1:]:
        parts = line.split()
        assert (len(parts) >= 2)
        try:
            value = int(parts[0])
        except ValueError:
            continue
        assert (value == len(classes))
        parts[-1] = parts[-1][0:-1]  # remove trailing comma
        classes.append(' '.join(parts[1:]))
    return classes


src/stactools/noaa_c_cap/constants.py
from pystac import Link
from pystac.provider import Provider, ProviderRole

COLLECTION_ID = "noaa-c-cap"
COLLECTION_TITLE = "C-CAP Regional Land Cover and Change"
COLLECTION_KEYWORDS = ["landcover", "coastal", "change"]
COLLECTION_PROVIDERS = [
    Provider(
        name=(
            "National Oceanic and Atmospheric Administration, "
            "Office for Coastal Management"
        ),
        roles=[
            ProviderRole.LICENSOR,
            ProviderRole.PRODUCER,
            ProviderRole.PROCESSOR,
            ProviderRole.HOST,
        ],
        url="https://coast.noaa.gov/digitalcoast/data/ccapregional.html",
    )
]
COLLECTION_DESCRIPTION = """Nationally standardized, raster-based inventories of \
land cover for the coastal areas of the U.S.  Data are derived, through the \
Coastal Change Analysis Program, from the analysis of multiple dates of remotely \
sensed imagery.  Two file types are available: individual dates that supply a \
wall-to-wall map, and change files that compare one date to another.  The use of \
standardized data and procedures assures consistency through time and \
across geographies.  C-CAP data forms the coastal expression of the National \
Land Cover Database (NLCD) and the A-16 land cover theme of the National Spatial \
Data Infrastructure.  The data are updated every 5 years."""
COLLECTION_LINKS = [
    Link(
        rel="describedby",
        target="https://coast.noaa.gov/data/digitalcoast/pdf/ccap-product-page.pdf",
        title="Handout - C-CAP Products",
        media_type="application/pdf",
    ),
    Link(
        rel="describedby",
        target="https://coast.noaa.gov/data/digitalcoast/pdf/ccap-faq-regional.pdf",
        title="Frequently Asked Questions",
        media_type="application/pdf",
    ),
    Link(
        rel="describedby",
        target="https://coast.noaa.gov/data/digitalcoast/pdf/ccap-class-scheme-regional.pdf",
        title="C-CAP Classification Scheme and Class Definitions",
        media_type="application/pdf",
    ),
    Link(
        rel="license",
        target="https://coast.noaa.gov/digitalcoast/data/ccapregional.html",
        title="C-CAP Regional Land Cover and Change",
        media_type="text/html",
    ),
]
CLASSIFICATION_EXTENSION_HREF = (
    "https://stac-extensions.github.io/classification/v1.0.0/schema.json"
)
CLASSIFICATION_CLASSES = [
    {
        "value": 0,
        "description": "Background",
    },
    {
        "value": 1,
        "description": "Unclassified (Cloud, Shadow, etc)",
    },
    {
        "value": 2,
        "description": "High Intensity Developed",
        "color-hint": "FFFFFF",
    },
    {
        "value": 3,
        "description": "Medium Intensity Developed",
        "color-hint": "C1C1C1",
    },
    {
        "value": 4,
        "description": "Low Intensity Developed",
        "color-hint": "979797",
    },
    {
        "value": 5,
        "description": "Developed Open Space",
        "color-hint": "CCCA42",
    },
    {
        "value": 6,
        "description": "Cultivated Crops",
        "color-hint": "4E2409",
    },
    {
        "value": 7,
        "description": "Pasture/Hay",
        "color-hint": "BCA05B",
    },
    {
        "value": 8,
        "description": "Grassland/Herbaceous",
        "color-hint": "EABB8E",
    },
    {
        "value": 9,
        "description": "Deciduous Forest",
        "color-hint": "6FEC48",
    },
    {
        "value": 10,
        "description": "Evergreen Forest",
        "color-hint": "15380A",
    },
    {
        "value": 11,
        "description": "Mixed Forest",
        "color-hint": "489D48",
    },
    {
        "value": 12,
        "description": "Scrub/Shrub",
        "color-hint": "6D6C1F",
    },
    {
        "value": 13,
        "description": "Palustrine Forested Wetland",
        "color-hint": "245A5A",
    },
    {
        "value": 14,
        "description": "Palustrine Scrub/Shrub Wetland",
        "color-hint": "E2742D",
    },
    {
        "value": 15,
        "description": "Palustrine Emergent Wetland (Persistent)",
        "color-hint": "DE3BEA",
    },
    {
        "value": 16,
        "description": "Estuarine Forested Wetland",
        "color-hint": "37083B",
    },
    {
        "value": 17,
        "description": "Estuarine Scrub/Shrub Wetland",
        "color-hint": "631569",
    },
    {
        "value": 18,
        "description": "Estuarine Emergent Wetland",
        "color-hint": "A028A9",
    },
    {
        "value": 19,
        "description": "Unconsolidated Shore",
        "color-hint": "6DEEF0",
    },
    {
        "value": 20,
        "description": "Bare Land",
        "color-hint": "F3F050",
    },
    {
        "value": 21,
        "description": "Open Water",
        "color-hint": "000A72",
    },
    {
        "value": 22,
        "description": "Palustrine Aquatic Bed",
        "color-hint": "0021E8",
    },
    {
        "value": 23,
        "description": "Estuarine Aquatic Bed",
        "color-hint": "6777D4",
    },
    {
        "value": 24,
        "description": "Tundra",
        "color-hint": "F3D147",
    },
    {
        "value": 25,
        "description": "Perennial Ice/Snow",
        "color-hint": "BBF7EF",
    },
]
COLLECTION_CITATION = (
    "National Oceanic and Atmospheric Administration, "
    "Office for Coastal Management. "
    "Name of Data Set: Coastal Change Analysis Program (C-CAP) Regional Land Cover. "
    "Charleston, SC: NOAA Office for Coastal Management. "
    "Accessed April 2022 at www.coast.noaa.gov/htdata/raster1/landcover/bulkdownload/30m_lc/."
)
SPATIAL_RESOLUTION = 30
